File: ASLRDisabler/Sources/ASLRDisabler/ASLRDisabler.m
//
//  ASLRDisabler.m
//  Clutch
//
//  Created by NinjaLikesCheez on 27/09/2021.
//  Copyright © 2021 Kim-Jong Cracks. All rights reserved.
//

#import "ASLRDisabler.h"

// #import "mach_vm.h"
#import <mach/mach_vm.h>
#import <mach/mach_init.h>
#import <mach/vm_map.h>

@import MachO.loader;

@implementation ASLRDisabler

// TODO: use the logger class when you can figure out why it's not importing via the bridging headers
+ (mach_vm_address_t)slideForPID:(pid_t)pid {
	vm_map_t targetTask = 0;

	if (task_for_pid(mach_task_self(), pid, &targetTask)) {
		printf("[!] Can't execute task_for_pid! Do you have the right permissions/entitlements?");
		return 0;
	}

	kern_return_t kernelReturn = 0;
	vm_address_t address = 0;

	while (1) {
		struct mach_header header = {0};
		vm_size_t size = 0;
		uint32_t depth = 0;
		mach_vm_size_t bytesRead = 0;
		struct vm_region_submap_info_64 info = {0};
		mach_msg_type_number_t count = VM_REGION_SUBMAP_INFO_COUNT_64;

		if (vm_region_recurse_64(targetTask, &address, &size, &depth, (vm_region_info_t)&info, &count)) {
			break;
		}

		kernelReturn = mach_vm_read_overwrite(targetTask,
																					(mach_vm_address_t)address,
																					(mach_vm_size_t)sizeof(struct mach_header),
																					(mach_vm_address_t)&header,
																					&bytesRead);

		if (kernelReturn == KERN_SUCCESS && bytesRead == sizeof(struct mach_header)) {
			/* only one image with MH_EXECUTE filetype */
			if ((header.magic == MH_MAGIC || header.magic == MH_MAGIC_64) && header.filetype == MH_EXECUTE) {
				printf("[+] Found main binary mach-o image @ %p!\n", (void *)address);
				return address;
			}
		}

		address += size;
	}

	printf("[!] Should not reach here");
	return 0;
}

@end

File: Clutch/Dumper/ASLRDisabler.swift
// swiftlint:disable comment_spacing
//////
//////  ASLRDisabler.swift
//////  Clutch
//////
//////  Created by NinjaLikesCheez Hedderwick on 12/11/2019.
//////  Copyright © 2019 Kim-Jong Cracks. All rights reserved.
//////
////
////import Foundation
////
//import MachO.loader
//
//class ASLRDisabler {
//	static func slide(forPID pid: Int32) -> mach_vm_address_t {
//		var target: vm_map_t = 0
//		var return_value: kern_return_t = 0 // don't remove this unused variable, it will break the function???
//
//		var kernel_return = task_for_pid(mach_task_self_, pid, &target)
//		if kernel_return != KERN_SUCCESS {
//			print("tfp0 failed: \(kernel_return)")
//			return 0
//		}
//
//		let VM_REGION_SUBMAP_INFO_V2_COUNT_64: mach_msg_type_number_t = mach_msg_type_number_t((MemoryLayout<vm_region_submap_info_data_64_t>.size / MemoryLayout<natural_t>.size))
//
//		var iter: vm_address_t = 0
//		var i: Int32 = 0
//
//		while true {
//			var address: vm_address_t = iter
//			var size: vm_size_t = 0
//			var depth: UInt32 = 0
//			var count: mach_msg_type_number_t = VM_REGION_SUBMAP_INFO_V2_COUNT_64
//
//			print("\n\n\n")
//			print("target(\(type(of: target))) = \(target)")
//			print("address(\(type(of: address))) = \(address)")
//			print("size(\(type(of: size))) = \(size)")
//			print("depth(\(type(of: depth))) = \(depth)")
//			print("count(\(type(of: count))) = \(count)")
//
//			kernel_return = withUnsafeMutablePointer(to: &i) {
//				vm_region_recurse_64(target, &address, &size, &depth, $0, &count)
//			}
//
//			if kernel_return != KERN_SUCCESS {
//				if let errorChar = mach_error_string(kernel_return),
//					 let errorString = String(utf8String: errorChar) {
//					print("vm_region_recurse_64 error: \(errorString)")
//				} else {
//					print("vm_region_recurse_64 error: \(kernel_return)")
//				}
//
//				break
//			}
//
//			var bytes_read: mach_vm_size_t = 0
//			var mh: mach_header = mach_header()
//
//			withUnsafePointer(to: &mh, {
//				let voidPtr = UnsafeRawPointer($0)
//				kernel_return = mach_vm_read_overwrite(
//					target,
//					mach_vm_address_t(address),
//					mach_vm_size_t(MemoryLayout<mach_header>.size),
//					mach_vm_address_t(voidPtr),
//					&bytes_read
//				)
//			})
//
//			if kernel_return != KERN_SUCCESS {
//				print("mach_vm_read_overwrite return: \(kernel_return)")
//			}
//
//			if kernel_return == KERN_SUCCESS && bytes_read == MemoryLayout<mach_header>.size {
//				if (mh.magic == MH_MAGIC || mh.magic == MH_MAGIC_64) && mh.filetype == MH_EXECUTE {
//					print("Found main binary mach-o image @ 0x\(address.hexString)")
//					return mach_vm_address_t(address)
//				}
//			}
//
//			iter = address + size
//		}
//
//		print("should not reach here!")
//		return 0
//	}
//}
// swiftlint:enable comment_spacing
